crates/ethereum/src/lib.rs

```diff
@@ -1,356 +1,252 @@
+        let (finalized_block_tx, mut finalized_block_rx) =
+            tokio::sync::mpsc::channel::<BlockNumber>(1);
+        tokio::spawn(async move {
+            let mut interval = tokio::time::interval(poll_interval);
+            loop {
+                interval.tick().await;
+                if let Ok(Some(finalized_block)) = provider_clone
+                    .get_block_by_number(BlockNumberOrTag::Finalized, false)
+                    .await
+                {
+                    let block_number = BlockNumber::new_or_panic(finalized_block.header.number);
+                    let _ = finalized_block_tx.send(block_number).await.unwrap();
+                }
+            }
         });
 
-        let url = Url::parse(&server.url("/"))?;
-        let eth = EthereumClient::new(url)?;
-        let chain_id = eth.get_chain().await?;
-
-        mock.assert();
-        assert_eq!(chain_id, EthereumChain::Mainnet);
-        Ok(())
+        // Add the finalized block stream to the select! macro
+        loop {
+            select! {
+                Some(state_update) = state_updates.next() => {
+                    // Decode the state update
+                    let state_update: Log<StarknetCoreContract::LogStateUpdate> = state_update.log_decode()?;
+                    let block_number = get_block_number(state_update.inner.blockNumber);
+                    // Add or remove to/from pending state updates accordingly
+                    if !state_update.removed {
+                    let state_update = EthereumStateUpdate {
+                            block_number,
+                        block_hash: get_block_hash(state_update.inner.blockHash),
+                        state_root: get_state_root(state_update.inner.globalRoot),
+                    };
+                    self.pending_state_updates.push(state_update);
+                    } else {
+                        self.pending_state_updates.retain(|update| update.block_number != block_number);
+                    }
+                }
+                Some(log) = logs.next() => {
+                    // Decode the message
+                    let log: Log<StarknetCoreContract::LogMessageToL2> = log.log_decode()?;
+                    // Create L1ToL2MessageHash from the log data
+                    let msg = L1ToL2MessageLog {
+                        message_hash: H256::from(log.inner.message_hash().to_be_bytes()),
+                        l1_tx_hash: log.transaction_hash.map(|hash| H256::from(hash.0)).unwrap_or_default(),
+                    };
+                    // Emit the message log
+                    callback(EthereumEvent::MessageLog(msg)).await;
+                }
+                Some(block_number) = finalized_block_rx.recv() => {
+                    // Drain all state updates up to (and including) the finalized block
+                    let idx = self.pending_state_updates.partition_point(|update| update.block_number <= block_number);
```

Comment:
I think we're mixing Ethereum block numbers and Starknet block numbers here.

The finalized block number is an _Ethereum_ block number, while `block_number` in `EthereumStateUpdate` is a _Starknet_ block number.